graphql_compiler/schema_transformation/merge_schemas.py

```diff
@@ -0,0 +1,173 @@
+                new_query_type_fields = new_definition.fields  # List[FieldDefinition]
+
+            elif isinstance(new_definition, ast_types.DirectiveDefinition):
+                if new_name in directives:  # existing directive
+                    if new_definition == directives[new_name]:  # definitions agree
+                        continue
+                    else:  # definitions disagree
+                        raise SchemaNameConflictError(
+                            u'Directive "{}" with definition "{}" has already been defined with '
+                            u'definition "{}".'.format(new_name, print_ast(new_definition),
+                                                       print_ast(directives[new_name]))
+                        )
+                # new directive
+                merged_schema_ast.definitions.append(new_definition)  # Add to AST
+                directives[new_name] = new_definition
+
+            elif isinstance(new_definition, ast_types.ScalarTypeDefinition):
+                if new_name in scalars:  # existing scalar
+                    continue
+                if new_name in name_to_schema_id:  # new scalar clashing with existing type
+                    raise SchemaNameConflictError(
+                        u'New scalar "{}" clashes with existing type.'.format(new_name)
+                    )
+                # new, valid scalar
+                merged_schema_ast.definitions.append(new_definition)  # Add to AST
+                scalars.add(new_name)
+
+            else:  # Generic type definition
+                if new_name in scalars:
+                    raise SchemaNameConflictError(
+                        u'New type "{}" clashes with existing scalar.'.format(new_name)
+                    )
+                if new_name in name_to_schema_id:
+                    raise SchemaNameConflictError(
+                        u'New type "{}" clashes with existing type.'.format(new_name)
+                    )
+                merged_schema_ast.definitions.append(new_definition)
+                name_to_schema_id[new_name] = cur_schema_id
+
+        # Concatenate all query type fields
+        # Since query_type was taken from the schema built from the input AST, the query type
+        # should never be not found
+        if new_query_type_fields is None:
+            raise AssertionError(u'Query type "{}" field definitions unexpected not '
+                                 u'found.'.format(cur_query_type))
+        # Note that as field names and type names have been confirmed to match up, and types
+        # were merged without name conflicts, query type fields can also be safely merged
+        query_type_index = 1  # Query type is the second entry in the list of definitions
+        merged_schema_ast.definitions[query_type_index].fields.extend(new_query_type_fields)
+
+    return MergedSchemaDescriptor(
+        schema_ast=merged_schema_ast,
+        name_to_schema_id=name_to_schema_id
+    )
```

Comment:
Would it be possible to refactor this function a bit? There's a lot of complexity here, and it's quite long. I'm finding it difficult to keep track of everything going on in here.